admiral/pkg/clusters/envoyfilter.go

```diff
@@ -0,0 +1,183 @@
+package clusters
+
+import (
+	"errors"
+	"fmt"
+	"github.com/gogo/protobuf/types"
+	v1 "github.com/istio-ecosystem/admiral/admiral/pkg/apis/admiral/v1"
+	"github.com/istio-ecosystem/admiral/admiral/pkg/controller/admiral"
+	"github.com/istio-ecosystem/admiral/admiral/pkg/controller/common"
+	log "github.com/sirupsen/logrus"
+	"istio.io/api/networking/v1alpha3"
+	networking "istio.io/client-go/pkg/apis/networking/v1alpha3"
+	metaV1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"strings"
+)
+
+var (
+	getSha1 = common.GetSha1
+)
+const hostsKey = "hosts: "
+
+func createOrUpdateEnvoyFilter( rc *RemoteController, routingPolicy *v1.RoutingPolicy, eventType admiral.EventType, workloadIdentityKey string, admiralCache *AdmiralCache, workloadSelectorMap map[string]string) (*networking.EnvoyFilter, error) {
+
+	envoyfilterSpec := constructEnvoyFilterStruct(routingPolicy, workloadSelectorMap)
+
+	selectorLabelsSha, err := getSha1(workloadIdentityKey+common.GetRoutingPolicyEnv(routingPolicy))
+	if err != nil {
+		log.Error("error ocurred while computing workload labels sha1")
+		return nil, err
+	}
+	if len(common.GetEnvoyFilterVersion()) == 0 {
+		log.Error("envoy filter version not supplied")
+		return nil, errors.New("envoy filter version not supplied")
+	}
+	envoyFilterName := fmt.Sprintf("%s-dynamicrouting-%s-%s", strings.ToLower(routingPolicy.Spec.Plugin), selectorLabelsSha, common.GetEnvoyFilterVersion())
+	envoyfilter := &networking.EnvoyFilter{
+		TypeMeta: metaV1.TypeMeta{
+			Kind:       "EnvoyFilter",
+			APIVersion: "networking.istio.io/v1alpha3",
+		},
+		ObjectMeta: metaV1.ObjectMeta{
+			Name:      envoyFilterName,
+			Namespace: common.NamespaceIstioSystem,
+		},
+		Spec: *envoyfilterSpec,
+	}
+
+	admiralCache.RoutingPolicyFilterCache.Put(workloadIdentityKey+common.GetRoutingPolicyEnv(routingPolicy), rc.ClusterID, envoyFilterName)
+	var filter *networking.EnvoyFilter
+	//get the envoyfilter if it exists. If it exists, update it. Otherwise create it.
+	if eventType == admiral.Add || eventType == admiral.Update {
+		// We query the API server instead of getting it from cache because there could be potential condition where the filter exists in the cache but not on the cluster.
+		filter, err = rc.RoutingPolicyController.IstioClient.NetworkingV1alpha3().EnvoyFilters(common.NamespaceIstioSystem).Get(envoyFilterName, metaV1.GetOptions{})
+		if err != nil {
+			log.Infof("msg=%s filtername=%s clustername=%s", "creating the envoy filter", envoyFilterName, rc.ClusterID)
```

Comment:
should we use https://github.com/kubernetes/apimachinery/blob/117bd9b56ec3fe6b4891d86097971e49bf9f68f4/pkg/api/errors/errors.go#L527 to ensure that it is not there? I think your intention is to call Create irrespective, as best effort. Which is good too.

Reply:
yup that was the idea let me test a little more in pre-prod to check if I can catch issues with always create